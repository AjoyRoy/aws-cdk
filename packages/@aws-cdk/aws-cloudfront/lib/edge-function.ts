import * as crypto from 'crypto';
import * as path from 'path';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as ssm from '@aws-cdk/aws-ssm';
import {
  BootstraplessSynthesizer, CfnResource, Construct as CoreConstruct, ConstructNode,
  CustomResource, CustomResourceProvider, CustomResourceProviderRuntime,
  DefaultStackSynthesizer, IStackSynthesizer, Resource, Stack, StackProps, Stage, Token,
} from '@aws-cdk/core';
import { Construct } from 'constructs';


/**
 * Properties for creating a Lambda@Edge function
 * @experimental
 */
export interface EdgeFunctionExperimentalProps extends lambda.FunctionProps { }

/**
 * A Lambda@Edge function.
 *
 * Convenience resource for requesting a Lambda function in the 'us-east-1' region for use with Lambda@Edge.
 * Implements several restrictions enforced by Lambda@Edge.
 *
 * @resource AWS::Lambda::Function
 * @experimental
 */
export class EdgeFunctionExperimental extends Resource implements lambda.IVersion {

  private static readonly EDGE_REGION: string = 'us-east-1';

  public readonly edgeArn: string;
  public readonly functionName: string;
  public readonly functionArn: string;
  public readonly grantPrincipal: iam.IPrincipal;
  public readonly isBoundToVpc = false;
  public readonly lambda: lambda.IFunction;
  public readonly permissionsNode: ConstructNode;
  public readonly role?: iam.IRole;
  public readonly version: string;

  // functionStack needed for `addAlias`.
  private readonly functionStack: Stack;
  private readonly edgeFunction: lambda.Function;

  constructor(scope: Construct, id: string, props: EdgeFunctionExperimentalProps) {
    super(scope, id);

    // Create a simple Function if we're already in us-east-1; otherwise create a cross-region stack.
    const regionIsUsEast1 = !Token.isUnresolved(this.stack.region) && this.stack.region === 'us-east-1';
    const { functionStack, edgeFunction, edgeArn } = regionIsUsEast1
      ? this.createInRegionFunction(id, props)
      : this.createCrossRegionFunction(id, props);

    this.functionStack = functionStack;
    this.edgeFunction = edgeFunction;
    this.edgeArn = edgeArn;

    this.functionArn = edgeArn;
    this.lambda = edgeFunction;
    this.functionName = this.lambda.functionName;
    this.grantPrincipal = this.lambda.role!;
    this.permissionsNode = this.lambda.permissionsNode;
    this.version = lambda.extractQualifierFromArn(this.functionArn);
  }

  /**
   * Returns 'this'. Convenience method to make `EdgeFunction` conform to the same interface as `Function`.
   */
  public get currentVersion(): lambda.IVersion {
    return this;
  }

  public addAlias(aliasName: string, options: lambda.AliasOptions = {}): lambda.Alias {
    return new lambda.Alias(this.functionStack, `Alias${aliasName}`, {
      aliasName,
      version: this.edgeFunction.currentVersion,
      ...options,
    });
  }

  /**
   * Not supported. Connections are only applicable to VPC-enabled functions.
   */
  public get connections(): ec2.Connections {
    throw new Error('Lambda@Edge does not support connections');
  }
  public get latestVersion(): lambda.IVersion {
    throw new Error('$LATEST function version cannot be used for Lambda@Edge');
  }

  public addEventSourceMapping(id: string, options: lambda.EventSourceMappingOptions): lambda.EventSourceMapping {
    return this.lambda.addEventSourceMapping(id, options);
  }
  public addPermission(id: string, permission: lambda.Permission): void {
    return this.lambda.addPermission(id, permission);
  }
  public addToRolePolicy(statement: iam.PolicyStatement): void {
    return this.lambda.addToRolePolicy(statement);
  }
  public grantInvoke(identity: iam.IGrantable): iam.Grant {
    return this.lambda.grantInvoke(identity);
  }
  public metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.lambda.metric(metricName, { ...props, region: EdgeFunctionExperimental.EDGE_REGION });
  }
  public metricDuration(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.lambda.metricDuration({ ...props, region: EdgeFunctionExperimental.EDGE_REGION });
  }
  public metricErrors(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.lambda.metricErrors({ ...props, region: EdgeFunctionExperimental.EDGE_REGION });
  }
  public metricInvocations(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.lambda.metricInvocations({ ...props, region: EdgeFunctionExperimental.EDGE_REGION });
  }
  public metricThrottles(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.lambda.metricThrottles({ ...props, region: EdgeFunctionExperimental.EDGE_REGION });
  }
  public addEventSource(source: lambda.IEventSource): void {
    return this.lambda.addEventSource(source);
  }
  public configureAsyncInvoke(options: lambda.EventInvokeConfigOptions): void {
    return this.lambda.configureAsyncInvoke(options);
  }

  /** Create a function in-region */
  private createInRegionFunction(id: string, props: lambda.FunctionProps): FunctionConfig {
    const role = props.role ?? defaultLambdaRole(this, id);
    const edgeFunction = new lambda.Function(this, 'Fn', {
      ...props,
      role,
    });

    return { edgeFunction, edgeArn: edgeFunction.currentVersion.edgeArn, functionStack: this.stack };
  }

  /** Create a support stack and function in us-east-1, and a SSM reader in-region */
  private createCrossRegionFunction(id: string, props: lambda.FunctionProps): FunctionConfig {
    const parameterNamePrefix = 'EdgeFunctionArn';
    const parameterName = `${parameterNamePrefix}${id}`;
    const functionStack = this.edgeStack();
    this.stack.addDependency(functionStack);

    const edgeFunction = functionStack.addEdgeFunction(id, parameterName, props);
    // This is used to determine when the function has changed, to refresh the ARN from the custom resource.
    const edgeFunctionHash = calculateFunctionHash(edgeFunction);

    // Prefix of the parameter ARN that applies to all EdgeFunctions.
    // This is necessary because the `CustomResourceProvider` is a singleton, and the `policyStatement`
    // must work for multiple EdgeFunctions.
    const parameterArnPrefix = this.stack.formatArn({
      service: 'ssm',
      region: EdgeFunctionExperimental.EDGE_REGION,
      resource: 'parameter',
      resourceName: parameterNamePrefix + '*',
    });

    const resourceType = 'Custom::CrossRegionStringParameterReader';
    const serviceToken = CustomResourceProvider.getOrCreate(this, resourceType, {
      codeDirectory: path.join(__dirname, 'edge-function'),
      runtime: CustomResourceProviderRuntime.NODEJS_12,
      policyStatements: [{
        Effect: 'Allow',
        Resource: parameterArnPrefix,
        Action: ['ssm:GetParameter'],
      }],
    });
    const resource = new CustomResource(this, 'ArnReader', {
      resourceType: resourceType,
      serviceToken,
      properties: {
        Region: EdgeFunctionExperimental.EDGE_REGION,
        ParameterName: parameterName,
        RefreshToken: edgeFunctionHash,
      },
    });
    const edgeArn = resource.getAttString('FunctionArn');

    return { edgeFunction, edgeArn, functionStack };
  }

  private edgeStack(): CrossRegionLambdaStack {
    const stage = this.node.root;
    if (!stage || !Stage.isStage(stage)) {
      throw new Error('stacks which use EdgeFunctions must be part of a CDK app or stage');
    }
    const region = this.env.region;
    if (Token.isUnresolved(region)) {
      throw new Error('stacks which use EdgeFunctions must have an explicitly set region');
    }

    const edgeStackId = `edge-lambda-stack-${region}`;
    let edgeStack = stage.node.tryFindChild(edgeStackId) as CrossRegionLambdaStack;
    if (!edgeStack) {
      edgeStack = new CrossRegionLambdaStack(stage, edgeStackId, {
        synthesizer: this.crossRegionSupportSynthesizer(),
        env: { region: EdgeFunctionExperimental.EDGE_REGION },
      });
    }
    return edgeStack;
  }

  // Stolen from `@aws-cdk/aws-codepipeline`'s `Pipeline`.
  private crossRegionSupportSynthesizer(): IStackSynthesizer | undefined {
    // If we have the new synthesizer we need a bootstrapless copy of it,
    // because we don't want to require bootstrapping the environment
    // of the account in this replication region.
    // Otherwise, return undefined to use the default.
    return (this.stack.synthesizer instanceof DefaultStackSynthesizer)
      ? new BootstraplessSynthesizer({
        deployRoleArn: this.stack.synthesizer.deployRoleArn,
        cloudFormationExecutionRoleArn: this.stack.synthesizer.cloudFormationExecutionRoleArn,
      })
      : undefined;
  }
}

/** Result of creating an in-region or cross-region function */
interface FunctionConfig {
  readonly edgeFunction: lambda.Function;
  readonly edgeArn: string;
  readonly functionStack: Stack;
}

class CrossRegionLambdaStack extends Stack {

  constructor(scope: CoreConstruct, id: string, props: StackProps) {
    super(scope, id, props);
  }

  public addEdgeFunction(id: string, parameterName: string, props: lambda.FunctionProps) {
    const role = props.role ?? defaultLambdaRole(this, id);

    const edgeFunction = new lambda.Function(this, id, {
      ...props,
      role,
    });

    new ssm.StringParameter(edgeFunction, 'Parameter', {
      parameterName,
      stringValue: edgeFunction.currentVersion.edgeArn,
    });

    return edgeFunction;
  }
}

function defaultLambdaRole(scope: Construct, id: string): iam.IRole {
  return new iam.Role(scope, `${id}ServiceRole`, {
    assumedBy: new iam.CompositePrincipal(
      new iam.ServicePrincipal('lambda.amazonaws.com'),
      new iam.ServicePrincipal('edgelambda.amazonaws.com'),
    ),
    managedPolicies: [iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')],
  });
}

// Stolen from @aws-lambda/lib/function-hash.ts, which isn't currently exported.
function calculateFunctionHash(fn: lambda.Function) {
  const stack = Stack.of(fn);
  const functionResource = fn.node.defaultChild as CfnResource;
  // render the cloudformation resource from this function
  const config = stack.resolve((functionResource as any)._toCloudFormation());

  const hash = crypto.createHash('md5');
  hash.update(JSON.stringify(config));
  return hash.digest('hex');
}
