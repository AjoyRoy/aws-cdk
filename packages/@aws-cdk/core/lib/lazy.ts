import { CDK_DEBUG, debugModeEnabled } from './debug';
import { IResolvable, IResolveContext } from './resolvable';
import { captureStackTrace } from './stack-trace';
import { Token } from './token';

/**
 * Interface for lazy string producers
 */
export interface IStringProducer {
  /**
   * Produce the string value
   */
  produce(context: IResolveContext): string | undefined;
}

/**
 * Interface for (stable) lazy string producers
 */
export interface IStableStringProducer {
  /**
   * Produce the string value
   */
  produce(): string | undefined;
}

/**
 * Interface for lazy list producers
 */
export interface IListProducer {
  /**
   * Produce the list value
   */
  produce(context: IResolveContext): string[] | undefined;
}

/**
 * Interface for (stable) lazy list producers
 */
export interface IStableListProducer {
  /**
   * Produce the list value
   */
  produce(): string[] | undefined;
}

/**
 * Interface for lazy number producers
 */
export interface INumberProducer {
  /**
   * Produce the number value
   */
  produce(context: IResolveContext): number | undefined;
}

/**
 * Interface for (stable) lazy number producers
 */
export interface IStableNumberProducer {
  /**
   * Produce the number value
   */
  produce(): number | undefined;
}

/**
 * Interface for lazy untyped value producers
 */
export interface IAnyProducer {
  /**
   * Produce the value
   */
  produce(context: IResolveContext): any;
}

/**
 * Interface for (stable) lazy untyped value producers
 */
export interface IStableAnyProducer {
  /**
   * Produce the value
   */
  produce(): any;
}

/**
 * Options for creating a lazy string token
 */
export interface LazyStringValueOptions {
  /**
   * Use the given name as a display hint
   *
   * @default - No hint
   */
  readonly displayHint?: string;
}

/**
 * Options for creating a lazy list token
 */
export interface LazyListValueOptions {
  /**
   * Use the given name as a display hint
   *
   * @default - No hint
   */
  readonly displayHint?: string;

  /**
   * If the produced list is empty, return 'undefined' instead
   *
   * @default false
   */
  readonly omitEmpty?: boolean;
}

/**
 * Options for creating lazy untyped tokens
 */
export interface LazyAnyValueOptions {
  /**
   * Use the given name as a display hint
   *
   * @default - No hint
   */
  readonly displayHint?: string;

  /**
   * If the produced value is an array and it is empty, return 'undefined' instead
   *
   * @default false
   */
  readonly omitEmptyArray?: boolean;
}

/**
 * Lazily produce a value
 *
 * Can be used to return a string, list or numeric value whose actual value
 * will only be calculated later, during synthesis.
 */
export class Lazy {
  /**
   * Defer the calculation of a string value to synthesis time
   *
   * Use this if you want to render a string to a template whose actual value depends on
   * some state mutation that may happen after the construct has been created.
   *
   * If you are simply looking to force a value to a `string` type and don't need
   * the calculation to be deferred, use `Token.asString()` instead.
   *
   * @deprecated Use `Lazy.string()` or `Lazy.uncachedString()` instead.
   */
  public static stringValue(producer: IStringProducer, options: LazyStringValueOptions = {}) {
    return Token.asString(new LazyString(producer, false), options);
  }

  /**
   * Defer the one-time calculation of a string value to synthesis time
   *
   * Use this if you want to render a string to a template whose actual value depends on
   * some state mutation that may happen after the construct has been created.
   *
   * If you are simply looking to force a value to a `string` type and don't need
   * the calculation to be deferred, use `Token.asString()` instead.
   *
   * The inner function will only be invoked one time and cannot depend on
   * resolution context.
   */
  public static string(producer: IStableStringProducer, options: LazyStringValueOptions = {}) {
    return Token.asString(new LazyString(producer, true), options);
  }

  /**
   * Defer the calculation of a string value to synthesis time
   *
   * Use if this function is not recommended; unless you know you need it for sure, you
   * probably don't. Use `Lazy.string()` instead.
   *
   * The inner function may be invoked multiple times during synthesis. You should
   * only use this function if the calculated value may change during the 'prepare'
   * phase of synthesis.
   */
  public static uncachedString(producer: IStringProducer, options: LazyStringValueOptions = {}) {
    return Token.asString(new LazyString(producer, false), options);
  }

  /**
   * Defer the one-time calculation of a number value to synthesis time
   *
   * Use this if you want to render a number to a template whose actual value depends on
   * some state mutation that may happen after the construct has been created.
   *
   * If you are simply looking to force a value to a `number` type and don't need
   * the calculation to be deferred, use `Token.asNumber()` instead.
   *
   * @deprecated Use `Lazy.number()` or `Lazy.uncachedNumber()` instead.
   */
  public static numberValue(producer: INumberProducer) {
    return Token.asNumber(new LazyNumber(producer, false));
  }

  /**
   * Defer the one-time calculation of a number value to synthesis time
   *
   * Use this if you want to render a number to a template whose actual value depends on
   * some state mutation that may happen after the construct has been created.
   *
   * If you are simply looking to force a value to a `number` type and don't need
   * the calculation to be deferred, use `Token.asNumber()` instead.
   *
   * The inner function will only be invoked one time and cannot depend on
   * resolution context.
   */
  public static number(producer: IStableNumberProducer) {
    return Token.asNumber(new LazyNumber(producer, true));
  }

  /**
   * Defer the calculation of a number value to synthesis time
   *
   * Use if this function is not recommended; unless you know you need it for sure, you
   * probably don't. Use `Lazy.number()` instead.
   *
   * The inner function may be invoked multiple times during synthesis. You should
   * only use this function if the calculated value may change during the 'prepare'
   * phase of synthesis.
   */
  public static uncachedNumber(producer: INumberProducer) {
    return Token.asNumber(new LazyNumber(producer, false));
  }

  /**
   * Defer the one-time calculation of a list value to synthesis time
   *
   * Use this if you want to render a list to a template whose actual value depends on
   * some state mutation that may happen after the construct has been created.
   *
   * If you are simply looking to force a value to a `string[]` type and don't need
   * the calculation to be deferred, use `Token.asList()` instead.
   *
   * @deprecated Use `Lazy.list()` or `Lazy.uncachedList()` instead.
   */
  public static listValue(producer: IListProducer, options: LazyListValueOptions = {}) {
    return Token.asList(new LazyList(producer, false, options), options);
  }

  /**
   * Defer the calculation of a list value to synthesis time
   *
   * Use if this function is not recommended; unless you know you need it for sure, you
   * probably don't. Use `Lazy.list()` instead.
   *
   * The inner function may be invoked multiple times during synthesis. You should
   * only use this function if the calculated value may change during the 'prepare'
   * phase of synthesis.
   */
  public static uncachedList(producer: IListProducer, options: LazyListValueOptions = {}) {
    return Token.asList(new LazyList(producer, false, options), options);
  }

  /**
   * Defer the one-time calculation of a list value to synthesis time
   *
   * Use this if you want to render a list to a template whose actual value depends on
   * some state mutation that may happen after the construct has been created.
   *
   * If you are simply looking to force a value to a `string[]` type and don't need
   * the calculation to be deferred, use `Token.asList()` instead.
   *
   * The inner function will only be invoked one time and cannot depend on
   * resolution context.
   */
  public static list(producer: IStableListProducer, options: LazyListValueOptions = {}) {
    return Token.asList(new LazyList(producer, true, options), options);
  }

  /**
   * Defer the one-time calculation of an arbitrarily typed value to synthesis time
   *
   * Use this if you want to render an object to a template whose actual value depends on
   * some state mutation that may happen after the construct has been created.
   *
   * @deprecated Use `Lazy.any()` or `Lazy.uncachedAny()` instead.
   */
  public static anyValue(producer: IAnyProducer, options: LazyAnyValueOptions = {}): IResolvable {
    return new LazyAny(producer, false, options);
  }

  /**
   * Defer the one-time calculation of an arbitrarily typed value to synthesis time
   *
   * Use this if you want to render an object to a template whose actual value depends on
   * some state mutation that may happen after the construct has been created.
   *
   * The inner function will only be invoked one time and cannot depend on
   * resolution context.
   */
  public static any(producer: IStableAnyProducer, options: LazyAnyValueOptions = {}): IResolvable {
    return new LazyAny(producer, true, options);
  }

  /**
   * Defer the calculation of an untyped value to synthesis time
   *
   * Use if this function is not recommended; unless you know you need it for sure, you
   * probably don't. Use `Lazy.any()` instead.
   *
   * The inner function may be invoked multiple times during synthesis. You should
   * only use this function if the calculated value may change during the 'prepare'
   * phase of synthesis.
   */
  public static uncachedAny(producer: IAnyProducer, options: LazyAnyValueOptions = {}): IResolvable {
    return new LazyAny(producer, false, options);
  }

  private constructor() {
  }
}

abstract class LazyBase<A> implements IResolvable {
  public readonly creationStack: string[];
  protected _cached?: A;

  constructor() {
    // Stack trace capture is conditionned to `debugModeEnabled()`, because
    // lazies can be created in a fairly thrashy way, and the stack traces are
    // large and slow to obtain; but are mostly useful only when debugging a
    // resolution issue.
    this.creationStack = debugModeEnabled()
      ? captureStackTrace(this.constructor)
      : [`Execute again with ${CDK_DEBUG}=true to capture stack traces`];
  }

  public abstract resolve(context: IResolveContext): any;

  public toString() {
    return Token.asString(this);
  }

  /**
   * Turn this Token into JSON
   *
   * Called automatically when JSON.stringify() is called on a Token.
   */
  public toJSON(): any {
    return '<unresolved-lazy>';
  }

}

class LazyString extends LazyBase<string> {

  constructor(private readonly producer: IStringProducer, private readonly cache: boolean) {
    super();
  }

  public resolve(context: IResolveContext) {
    if (this.cache) {
      return this._cached ?? (this._cached = this.producer.produce(context));
    } else {
      return this.producer.produce(context);
    }
  }
}

class LazyNumber extends LazyBase<number> {
  constructor(private readonly producer: INumberProducer, private readonly cache: boolean) {
    super();
  }

  public resolve(context: IResolveContext) {
    if (this.cache) {
      return this._cached ?? (this._cached = this.producer.produce(context));
    } else {
      return this.producer.produce(context);
    }
  }
}

class LazyList extends LazyBase<Array<string>> {
  constructor(private readonly producer: IListProducer, private readonly cache: boolean, private readonly options: LazyListValueOptions = {}) {
    super();
  }

  public resolve(context: IResolveContext) {
    if (!this.cache || this._cached === undefined) {
      this._cached = this.producer.produce(context);
    }

    if (this._cached !== undefined && this._cached.length === 0 && this.options.omitEmpty) {
      return undefined;
    }
    return this._cached;
  }
}

class LazyAny extends LazyBase<any> {
  constructor(private readonly producer: IAnyProducer, private readonly cache: boolean, private readonly options: LazyAnyValueOptions = {}) {
    super();
  }

  public resolve(context: IResolveContext) {
    if (!this.cache || this._cached === undefined) {
      this._cached = this.producer.produce(context);
    }

    if (Array.isArray(this._cached) && this._cached.length === 0 && this.options.omitEmptyArray) {
      return undefined;
    }
    return this._cached;
  }
}
