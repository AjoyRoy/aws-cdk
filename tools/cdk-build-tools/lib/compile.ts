import { makeExecutable, shell } from './os';
import { CDKBuildOptions, CompilerOverrides, currentPackageJson, isJsii, packageCompiler } from './package-info';
import { Timers } from './timer';

const DEFAULT_ROSETTA = require.resolve('jsii-rosetta/bin/jsii-rosetta');

/**
 * Run the compiler on the current package
 */
export async function compileCurrentPackage(options: CDKBuildOptions, timers: Timers, compilers: CompilerOverrides = {}): Promise<void> {
  const env = options.env;
  await shell(packageCompiler(compilers), { timers, env });

  if (isJsii() && options.rosetta === 'strict') {
    await shell([
      compilers.rosetta ?? DEFAULT_ROSETTA,
      'extract',
      '--compile',
      '--fail',
    ], { timers, env });
  }

  // Find files in bin/ that look like they should be executable, and make them so.
  const scripts = currentPackageJson().bin || {};
  for (const script of Object.values(scripts) as any) {
    await makeExecutable(script);
  }
}
