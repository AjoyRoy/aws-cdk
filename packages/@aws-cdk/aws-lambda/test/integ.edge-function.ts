/// !cdk-integ *
import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import * as lambda from '../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'EdgeFunction', {
  env: {
    region: 'eu-west-1', // Anything but us-east-1
  },
});

new lambda.EdgeFunction(stack, 'MyLambda', {
  code: lambda.Code.fromAsset(path.join(__dirname, 'my-lambda-handler')),
  handler: 'index.main',
  runtime: lambda.Runtime.PYTHON_3_6,
});

app.synth();
