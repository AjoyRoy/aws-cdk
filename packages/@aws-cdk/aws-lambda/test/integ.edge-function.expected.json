[
  {
    "Resources": {
      "MyLambdaArnReader2D7DDF01": {
        "Type": "Custom::CrossRegionStringParameterReader",
        "Properties": {
          "ServiceToken": {
            "Fn::GetAtt": [
              "CustomCrossRegionStringParameterReaderCustomResourceProviderHandler65B5F33A",
              "Arn"
            ]
          },
          "Region": "us-east-1",
          "ParameterName": "EdgeFunctionArnMyLambda",
          "RefreshToken": "f69585dd708dc595835ff1c75b0b01a6"
        },
        "UpdateReplacePolicy": "Delete",
        "DeletionPolicy": "Delete"
      },
      "CustomCrossRegionStringParameterReaderCustomResourceProviderRole71CD6825": {
        "Type": "AWS::IAM::Role",
        "Properties": {
          "AssumeRolePolicyDocument": {
            "Version": "2012-10-17",
            "Statement": [
              {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": {
                  "Service": "lambda.amazonaws.com"
                }
              }
            ]
          },
          "ManagedPolicyArns": [
            {
              "Fn::Sub": "arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
            }
          ],
          "Policies": [
            {
              "PolicyName": "Inline",
              "PolicyDocument": {
                "Version": "2012-10-17",
                "Statement": [
                  {
                    "Effect": "Allow",
                    "Resource": {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition"
                          },
                          ":ssm:us-east-1:",
                          {
                            "Ref": "AWS::AccountId"
                          },
                          ":parameter/EdgeFunctionArn*"
                        ]
                      ]
                    },
                    "Action": [
                      "ssm:GetParameter"
                    ]
                  }
                ]
              }
            }
          ]
        }
      },
      "CustomCrossRegionStringParameterReaderCustomResourceProviderHandler65B5F33A": {
        "Type": "AWS::Lambda::Function",
        "Properties": {
          "Code": {
            "S3Bucket": {
              "Ref": "AssetParameters45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241S3BucketF1BC72A7"
            },
            "S3Key": {
              "Fn::Join": [
                "",
                [
                  {
                    "Fn::Select": [
                      0,
                      {
                        "Fn::Split": [
                          "||",
                          {
                            "Ref": "AssetParameters45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241S3VersionKey7AD83AC7"
                          }
                        ]
                      }
                    ]
                  },
                  {
                    "Fn::Select": [
                      1,
                      {
                        "Fn::Split": [
                          "||",
                          {
                            "Ref": "AssetParameters45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241S3VersionKey7AD83AC7"
                          }
                        ]
                      }
                    ]
                  }
                ]
              ]
            }
          },
          "Timeout": 900,
          "MemorySize": 128,
          "Handler": "__entrypoint__.handler",
          "Role": {
            "Fn::GetAtt": [
              "CustomCrossRegionStringParameterReaderCustomResourceProviderRole71CD6825",
              "Arn"
            ]
          },
          "Runtime": "nodejs12.x"
        },
        "DependsOn": [
          "CustomCrossRegionStringParameterReaderCustomResourceProviderRole71CD6825"
        ]
      }
    },
    "Parameters": {
      "AssetParameters45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241S3BucketF1BC72A7": {
        "Type": "String",
        "Description": "S3 bucket for asset \"45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241\""
      },
      "AssetParameters45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241S3VersionKey7AD83AC7": {
        "Type": "String",
        "Description": "S3 key for asset version \"45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241\""
      },
      "AssetParameters45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241ArtifactHash1B3D1B80": {
        "Type": "String",
        "Description": "Artifact hash for asset \"45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241\""
      }
    }
  },
  {
    "Resources": {
      "MyLambdaServiceRoleC379DB2A": {
        "Type": "AWS::IAM::Role",
        "Properties": {
          "AssumeRolePolicyDocument": {
            "Statement": [
              {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": {
                  "Service": [
                    "lambda.amazonaws.com",
                    "edgelambda.amazonaws.com"
                  ]
                }
              }
            ],
            "Version": "2012-10-17"
          },
          "ManagedPolicyArns": [
            {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition"
                  },
                  ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
                ]
              ]
            }
          ]
        }
      },
      "MyLambdaCCE802FB": {
        "Type": "AWS::Lambda::Function",
        "Properties": {
          "Code": {
            "S3Bucket": {
              "Ref": "AssetParameters9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232S3Bucket1354C645"
            },
            "S3Key": {
              "Fn::Join": [
                "",
                [
                  {
                    "Fn::Select": [
                      0,
                      {
                        "Fn::Split": [
                          "||",
                          {
                            "Ref": "AssetParameters9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232S3VersionKey5D873FAC"
                          }
                        ]
                      }
                    ]
                  },
                  {
                    "Fn::Select": [
                      1,
                      {
                        "Fn::Split": [
                          "||",
                          {
                            "Ref": "AssetParameters9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232S3VersionKey5D873FAC"
                          }
                        ]
                      }
                    ]
                  }
                ]
              ]
            }
          },
          "Handler": "index.main",
          "Role": {
            "Fn::GetAtt": [
              "MyLambdaServiceRoleC379DB2A",
              "Arn"
            ]
          },
          "Runtime": "python3.6"
        },
        "DependsOn": [
          "MyLambdaServiceRoleC379DB2A"
        ]
      },
      "MyLambdaCurrentVersionE7A382CCfc7f4f33e2ce545541fb53c9ea4a3766": {
        "Type": "AWS::Lambda::Version",
        "Properties": {
          "FunctionName": {
            "Ref": "MyLambdaCCE802FB"
          }
        }
      },
      "MyLambdaParameter72075EB0": {
        "Type": "AWS::SSM::Parameter",
        "Properties": {
          "Type": "String",
          "Value": {
            "Ref": "MyLambdaCurrentVersionE7A382CCfc7f4f33e2ce545541fb53c9ea4a3766"
          },
          "Name": "EdgeFunctionArnMyLambda"
        }
      }
    },
    "Parameters": {
      "AssetParameters9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232S3Bucket1354C645": {
        "Type": "String",
        "Description": "S3 bucket for asset \"9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232\""
      },
      "AssetParameters9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232S3VersionKey5D873FAC": {
        "Type": "String",
        "Description": "S3 key for asset version \"9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232\""
      },
      "AssetParameters9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232ArtifactHashFA16AACF": {
        "Type": "String",
        "Description": "Artifact hash for asset \"9678c34eca93259d11f2d714177347afd66c50116e1e08996eff893d3ca81232\""
      }
    }
  }
]